### PR TITLE
JFG-921: Add JHttpResponseStatusValidator

### PR DESCRIPTION
**Pull request by anton-antonenko**

### CODE REVIEW

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/JHttpResponseStatusValidator.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import com.griddynamics.jagger.invoker.v2.JHttpResponse;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.collect.Sets.newHashSet;
+
+/**
+ * Validates JHttpResponse status code.<p>
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/JHttpResponseStatusValidator.java]: please remove \<p>

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/JHttpResponseStatusValidator.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * If {@link JHttpResponseStatusValidator#validStatusCodesRegexp} is not null, it will be used for validation, otherwise
+ * {@link JHttpResponseStatusValidator#validStatusCodes} is used.
+ */
+public class JHttpResponseStatusValidator extends ResponseValidator<JHttpQuery, JHttpEndpoint, JHttpResponse> {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/JHttpResponseStatusValidator.java]: WHYT if we will move validator to the same package like all JHttp classes?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Provides default response status validator {@link JHttpResponseStatusValidator}.
+ */
+public class ResponseStatusValidatorProvider implements ResponseValidatorProvider {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: let's make it also JHttp

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+     * @throws PatternSyntaxException If the expression's syntax is invalid
+     * @apiNote Usage: <pre>
+     * {@code
+     * ResponseStatusValidatorProvider validator = ResponseStatusValidatorProvider.withValidCodes("(200|201|202)");
+     * }
+     * <pre/>
+     */
+    public static ResponseStatusValidatorProvider withValidCodes(String stringRegExp) {
+        return new ResponseStatusValidatorProvider(Pattern.compile(stringRegExp));
+    }
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: what about withValidators(List<>)?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * @param codes valid codes
+     * @return this
+     * @apiNote Usage: <pre>
+     * {@code
+     * ResponseStatusValidatorProvider validator = ResponseStatusValidatorProvider.withValidCodes(200, 201, 202);
+     * }
+     * <pre/>
+     */
+    public static ResponseStatusValidatorProvider withValidCodes(Integer... codes) {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: we are breaking or approach: withSomething - optional parameters

Comment (anton-antonenko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: what do you mean?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java (hunk 4 above)
Comment (anton-antonenko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: By Validators you mean regexps?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java (hunk 5 above)
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: parameters passed in constructors are mandatory
parameters passed via setters - are optional

in this case, status codes are mandatory parameters, but they are passed via setters

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java (hunk 4 above)
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: it was  a typo. withValidCodes(List<>)

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java (hunk 5 above)
Comment (anton-antonenko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: It's not a builder - all these methods are factory methods.
And user must choose one of them to use. 
They cannot be combined.
We discussed it with Oleksiy and he asked me to replace builders with these factory methods.

Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: don't we have method 'of' for factories

Comment (anton-antonenko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/ResponseStatusValidatorProvider.java]: Yes, I can rename them for consistency

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/JHttpResponseStatusValidatorProvider.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+     * @param stringRegExps list of RegExps defining valid codes
+     * @return new JHttpResponseStatusValidatorProvider with valid codes regExps provided in stringRegExps
+     * @throws PatternSyntaxException If the expression's syntax is invalid
+     * @apiNote Usage: <pre>
+     * {@code
+     * JHttpResponseStatusValidatorProvider validator = JHttpResponseStatusValidatorProvider.of(newArrayList("(200|201|202)"));
+     * }
+     * <pre/>
+     */
+    public static JHttpResponseStatusValidatorProvider of(List<String> stringRegExps) {
```
Comment (anton-antonenko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/collector/JHttpResponseStatusValidatorProvider.java]: What for? There's already of(Integer... codes), which works the same

**Pull request merged**